eos/utils/kmatrix-impl.hh

```diff
@@ -0,0 +1,213 @@
+            throw InternalError("The size of the array of background constants does not match nchannels_.");
+        if (bkgcst.size() != 0 and bkgcst[0].size() != nchannels_)
+            throw InternalError("The array of background constants is not square.");
+        if (resonances.size() != nresonances_)
+            throw InternalError("The size of the resonances array does not match nresonances_.");
+
+        // Perform pointer checks
+        if (_Khat == nullptr)
+            throw std::bad_alloc();
+        if (_That == nullptr)
+            throw std::bad_alloc();
+        if (_tmp_1 == nullptr)
+            throw std::bad_alloc();
+        if (_tmp_2 == nullptr)
+            throw std::bad_alloc();
+        if (_perm == nullptr)
+            throw std::bad_alloc();
+    }
+
+    template <unsigned nchannels_, unsigned nresonances_>
+    KMatrix<nchannels_, nresonances_>::~KMatrix()
+    {
+        if (_perm)
+            gsl_permutation_free(_perm);
+        _perm = nullptr;
+
+        if (_tmp_1)
+            gsl_matrix_complex_free(_tmp_1);
+        _tmp_1 = nullptr;
+
+        if (_tmp_2)
+            gsl_matrix_complex_free(_tmp_2);
+        _tmp_2 = nullptr;
+
+        if (_That)
+            gsl_matrix_complex_free(_That);
+        _That = nullptr;
+
+        if (_Khat)
+            gsl_matrix_complex_free(_Khat);
+        _Khat = nullptr;
+    }
+
+
+    //Return row n° rowindex of the T matrix defined as T = (1-i*rho*K)^(-1)*K
+    template <unsigned nchannels_, unsigned nresonances_>
+    std::array<complex<double>, nchannels_>
+    KMatrix<nchannels_, nresonances_>::tmatrix_row(unsigned rowindex, double s) const
+    {
+
+        std::array<complex<double>, nchannels_> tmatrixrow;
+        auto channels = this->_channels;
+        auto resonances = this->_resonances;
+        auto bkgcst = this->_bkgcst;
+
+
+        ///////////////////
+        //1. Fill tmp1 = rho matrix
+        ///////////////////
+        gsl_matrix_complex_set_zero(_tmp_1);
+
+        for (size_t i = 0 ; i < nchannels_ ; ++i)
+            {
+            complex<double> rhoentry = channels[i]->rho(s);
+            gsl_matrix_complex_set(_tmp_1,  i,  i, gsl_complex_rect(rhoentry.real(), rhoentry.imag()));
+            }
+
+        ///////////////////
+        // 2. Fill Khat
+        ///////////////////
+        for (size_t i = 0 ; i < nchannels_ ; ++i)
+            {
+            for (size_t j = 0 ; j < nchannels_ ; ++j)
+                {
+                complex<double> entry = complex<double>(bkgcst[i][j].evaluate(), 0.0);
+
+                for (size_t a = 0 ; a < nresonances_ ; ++a)
+                    {
+
+                    double mres = resonances[a]->_m;
+
+                    Parameter g0rci = channels[i]->_g0s[a];
+                    Parameter g0rcj = channels[j]->_g0s[a];
+
+
+                    // Avoid the pole by adding a small imaginary part
+                    complex<double> mres2 = { mres * mres, 1.0e-7 };
```

Comment:
I would prefer if we did not use a magic number for the imaginary part. Can we use ```std::sqrt(std::numeric_limits<double>::min())``` or something like that?

Reply:
Problem solved as discussed